bpo-35813: Tests and docs for shared_memory

applio opened the pull request:
Tests and docs are now present.  Discussion around the API for the SharedMemory class and the appropriateness of the shareable_wrap() function is requested in particular.


<!-- issue-number: [bpo-35813](https://bugs.python.org/issue35813) -->
https://bugs.python.org/issue35813
<!-- /issue-number -->

applio committed:
Added tests for shared_memory submodule.

applio committed:
Added tests for ShareableList.

applio committed:
Fix bug in allocationn size during creation of empty ShareableList illuminated by existing test run on Linux.

applio committed:
Initial set of docs for shared_memory module.

applio committed:
Added docs for ShareableList, added doctree entry for shared_memory submodule, name refactoring for greater clarity.

applio committed:
Added examples to SharedMemoryManager docs, for ease of documentation switched away from exclusively registered functions to some explicit methods on SharedMemoryManager.

applio committed:
Wording tweaks to docs.

applio committed:
Fix test failures on Windows.

applio committed:
Added tests around SharedMemoryManager.

applio committed:
Documentation tweaks.

applio committed:
Fix inappropriate test on Windows.

applio committed:
Further documentation tweaks.

applio committed:
Fix bare exception.

eamanu commented on Doc/library/multiprocessing.shared_memory.rst:
```diff
@@ -0,0 +1,381 @@
+
+This module provides a class, :class:`SharedMemory`, for the allocation
+and management of shared memory to be accessed by one or more processes
+on a multicore or symmetric multiprocessor (SMP) machine.  To assist with
```
the double whitepace on `machine.  To assist` if for more readable?

applio replied:
In the resulting html, whether one or two spaces are in the source only one space is presented in the browser.  The convention of using two spaces predates patterns of using only one space after a period.  As far as I know, both are "correct".  Thankfully the output is the same either way.

applio replied:
I did not go looking for this but I accidentally stumbled across it in the dev guide (see devguide.python.org for lots more such goodies):
```
A sentence-ending period may be followed by one or two spaces;
while reST ignores the second space, it is customarily put in
by some users, for example to aid Emacs’ auto-fill mode.
```

eamanu replied:
Oh great thanks for the clarification. I had the idea that double space is used just on docstring.

eamanu commented on Doc/library/multiprocessing.shared_memory.rst:
```diff
@@ -0,0 +1,381 @@
+
+   >>> from multiprocessing import shared_memory
+   >>> shm_a = shared_memory.SharedMemory(None, size=10)
+   >>> type(shm_b.buf)
```
Where do you define `shm_b?`

applio replied:
Great spotting that missed detail!  Fixed in commit da7731db412a94d9f06a99500b31c9147e601e3e.

eamanu commented on Doc/library/multiprocessing.shared_memory.rst:
```diff
@@ -0,0 +1,381 @@
+   >>> len(buffer)
+   10
+   >>> buffer[:4] = bytearray([22, 33, 44, 55])  # Modify multiple at once
+   >>> buffer[4] = 100                           # Modify single byte at a time
```
I don't know if pep8 will fail here. Because it say you that you need have two whitespaces.

applio replied:
I believe this physical alignment of two related comments helps readability significantly.  To not do so would make the example much less readable.  Pep8 encourages considering this in particular.

eamanu commented on Doc/library/multiprocessing.shared_memory.rst:
```diff
@@ -0,0 +1,381 @@
+   >>> array.array('b', shm_b.buf[:5])  # Copy the data into a new array.array
+   array('b', [22, 33, 44, 55, 100])
+   >>> shm_b.buf[:5] = b'howdy'  # Modify via shm_b using bytes
+   >>> bytes(shm_a.buf[:5])      # Access via shm_a
```
same here

applio replied:
Same comment as on the prior; I believe this significantly helps readability in a situation where such help is needed.

eamanu commented on Doc/library/multiprocessing.shared_memory.rst:
```diff
@@ -0,0 +1,381 @@
+
+
+The following example demonstrates a practical use of the :class:`SharedMemory`
+class with `NumPy arrays <https://www.numpy.org/>`_, accessing the
```
I don't know if is necessary show a sample with NumPy, it's just an opinion

applio replied:
The use of NumPy arrays with shared memory is anticipated to be one of the most popular use cases for working with shared memory.  Demonstrating this combination is likely very important to a large number of people.

eamanu commented on Doc/library/multiprocessing.shared_memory.rst:
```diff
@@ -0,0 +1,381 @@
+
+   A call to :meth:`~multiprocessing.managers.BaseManager.start` on a
+   :class:`SharedMemoryManager` instance causes a new process to be started.
+   This new process's sole purpose is to manage the life cycle
```
`process'` ?

applio replied:
After double-checking with a technical editor just now, I have been assured that either is correct.

eamanu replied:
Great!

eamanu commented on Lib/multiprocessing/shared_memory.py:
```diff
@@ -8,34 +8,120 @@
 from functools import reduce
 import mmap
-from .managers import DictProxy, SyncManager, Server
+from .managers import dispatch, BaseManager, Server, State, ProcessError, \
```
What do you think about use `()` on imports? This way you avoid use `\`

applio replied:
I like both styles.  At the moment, I think it best to stay consistent with the pattern that already exists in multiprocessing and its submodules.

eamanu commented on Lib/multiprocessing/shared_memory.py:
```diff
@@ -8,34 +8,120 @@
+    import ctypes
+    from ctypes import wintypes
+
+    #kernel32 = ctypes.wintypes.WinDLL("kernel32", use_last_error=True)
```
```suggestion
    # kernel32 = ctypes.wintypes.WinDLL("kernel32", use_last_error=True)
```

applio replied:
Removed in da7731db412a94d9f06a99500b31c9147e601e3e.

eamanu reviewed (commented):
Good job!!! Please consider review my comments. Please review the CI problems.

applio committed:
Removed __copyright__.

applio committed:
Fixed typo in doc, removed comment.

applio committed:
Merge remote-tracking branch 'upstream/master' into enh-tests-shmem

giampaolo commented on Doc/library/multiprocessing.shared_memory.rst:
```diff
@@ -0,0 +1,381 @@
+   *name* is the unique name for the requested shared memory, specified as
+   a string.  If ``None`` is supplied for the name, a new shared memory
+   block with a novel name will be created without needing to also specify
+   ``flags``.
```
Since it appears `name` is optional, why not make it a kwarg?

applio replied:
Moved this topic to b.p.o. starting with msg335660 there.

giampaolo commented on Doc/library/multiprocessing.shared_memory.rst:
```diff
@@ -0,0 +1,381 @@
+   exception will be raised.  To request the creation of a new shared
+   memory block, set to ``O_CREX``.  To request the optional creation of a
+   new shared memory block or attach to an existing one by the same name,
+   set to ``O_CREAT``.
```
It appears deciding whether to use `O_CREX` / `O_CREAT` can be implemented as an internal detail based on whether the name is provided or not. Are there other use cases for `flags` argument?

applio replied:
Moved this topic to b.p.o. starting with msg335660 there.

giampaolo commented on Doc/library/multiprocessing.shared_memory.rst:
```diff
@@ -0,0 +1,381 @@
+   set to ``O_CREAT``.
+
+   *mode* controls user/group/all-based read/write permissions on the
+   shared memory block.  Its specification is not enforceable on all platforms.
```
Is `mode` intended to work like the mode of UNIX file? (I see it defaults to `0o600` so it appears it does). What is the use case for it?

applio replied:
Moved this topic to b.p.o. starting with msg335660 there.

giampaolo commented on Doc/library/multiprocessing.shared_memory.rst:
```diff
@@ -0,0 +1,381 @@
+copying of data.
+
+
+.. class:: SharedMemory(name, flags=None, mode=0o600, size=0, read_only=False)
```
If I understand your code samples it seems this class/object can be returned via the `SharedMemoryManager` as in:

```python
>>> mgr = SharedMemoryManager()
>>> mgr.start()
>>> raw_shm = smm.SharedMemory(size=128) 
``` 

As such, what's the use case for using `SharedMemory` directly? Also, what's the use case for the `name` parameter? I understand it you can use it for attaching to a pre-existing instance of `SharedMemory`, but this kind of diverges from the base `SyncManager` which tipically allows sharing objects by passing them as arguments for the worker. E.g., with a plain dict:

```python
import multiprocessing
import multiprocessing.managers

def worker(d):
    print(d)  # prints {'foo': 1}
    d['bar'] = 2  # I can also change the dict from here (child process)

manager = multiprocessing.managers.SyncManager()
manager.start()
d = manager.dict()
d['foo'] = 1
proc = multiprocessing.Process(target=worker, args=(d, ))
proc.start()
proc.join()
```

I wonder if this (the `name`) can be made transparent (aka an implementation detail).

applio replied:
Moved this topic to b.p.o. starting with msg335622 there.

giampaolo commented on Doc/library/multiprocessing.shared_memory.rst:
```diff
@@ -0,0 +1,381 @@
+
+   *read_only* controls whether a shared memory block is to be available
+   for only reading or for both reading and writing.  Its specification is
+   not enforceable on all platforms.
```
You should mention what platforms implement this and what happens when it's not and `True` is passed. Also, since this appears not portable across platform, perhaps it's better to not support it at all.

applio replied:
Agreed -- with the simpler API introduced in commit 0d3d06f7cbcb47b2dee6c027c35265795052c899, we no longer expose users to such complications.

applio committed:
Updated SharedMemoryManager preliminary tests to reflect change of not supporting all registered functions on SyncManager.

giampaolo commented on Doc/library/multiprocessing.shared_memory.rst:
```diff
@@ -0,0 +1,381 @@
+      inside the shared memory block after ``unlink()`` has been called may
+      result in memory access errors.  Note: the last process relinquishing
+      its hold on a shared memory block may call ``unlink()`` and
+      :meth:`close()` in either order.
```
Mmm... what's the point of having both `close` and `unlink`? Why not simply having `close()` also destroy the memory blocks?

applio replied:
After all instances have called `close()`, there are still reasons to potentially want to have a shared memory block persist.  An existing technique that some services use to preserve data even when the service dies unexpectedly is to store critical data in a shared memory block with an established name -- when the service restarts, it looks for the shared memory to still be there by name, thereby reducing the startup time of the service.  Other use cases likely helped motivate the mentality behind the Windows implementation of shared memory which always preserves the shared memory block until the last process with a handle on it terminates -- Windows offers no way to perform an `unlink()` action ahead of time.  Closest thing you can do on Windows is truncate its size to near zero but even that might not trigger a partial release of memory space.  Across platforms, the common mentality and conscious design is to preserve shared memory until its release is explicitly demanded.

giampaolo commented on Doc/library/multiprocessing.shared_memory.rst:
```diff
@@ -0,0 +1,381 @@
+   Because some platforms choose to allocate chunks of memory based upon
+   that platform's memory page size, the exact size of the shared memory
+   block may be larger or equal to the size requested.  When attaching to an
+   existing shared memory block, set to ``0`` (which is the default).
```
What happens if `size` is exceeded? `MemoryError`? If so it should be documented.
Also, this looks kind of a low level detail and I'm unsure how/when to use it or what it means exactly. E.g. say I create a a list of 10 elements and specify `size=5`. Is the total object size supposed to be 50 bytes?

applio replied:
When attaching to an existing shared memory block, supplying a `size` larger than the currently allocated size of the block will trigger an attempt by the OS kernel to grow the size of the shared memory block.  Existing handles on that same shared memory block will remain unaffected; they will still only have access to the first N bytes and their manner of accessing and referring to that shared memory does not need to change.  If the OS kernel is unable to grow the size of the shared memory block to the requested `size` for any reason, then the attempt to attach to the existing shared memory block will fail.

giampaolo replied:
Probably worth documenting IMO. Also, I wonder whether maybe this should be called `sizehint` instead? There are other parts of the lib where this is used, so just an idea...

applio replied:
Added to documentation in commit 34f1e9a7514553f1ee017137aacda8ee4246ee01.

giampaolo commented on Doc/library/multiprocessing.shared_memory.rst:
```diff
@@ -0,0 +1,381 @@
+   .. method:: Semaphore([value])
+
+      Create a shared :class:`threading.Semaphore` object and return a proxy
+      for it.
```
All these objects are already documented in the managers section:
https://docs.python.org/3/library/multiprocessing.html#managers
By moving `SharedMemoryManager` (and `SharedMemoryServer`) in `multiprocessing.managers` namespace we can avoid this repetion, and only document the objects/parts that differ (e.g. `ShareableList`).

applio replied:
Per my longer post to b.p.o. regarding this, I now think we should remove these from SharedMemoryManager.

giampaolo replied:
Even if you remove them it still remains a manager. It subclasses `BaseManager`, shares its interface (`get_server()`, `start()` and `shutdown()`) and you use it in the same way as `SyncManager`. So the more natural place to me seems `multiprocessing.managers` namespace. 

The remaining classes are more questionable, essentially because there's two ways of using this feauture, via the manager and directly by using "name"s to attach to other processes. 

~I think they definitively have to live in `shared_memory.py`, but there's doubt around which ones should be public/documented. Part of my confusion derives from the fact that this facility can be accessed in two ways: through the manager or via `SharedMemory`. Using the manager is easier because it exposes less details about the underlying implementation. ~~A question I [asked above](https://github.com/python/cpython/pull/11816/files#r255695988) is: why should the user want to use `SharedMemory` directly? Why not just go through the manager?~~ [EDIT: see [why](https://bugs.python.org/issue35813#msg335622)]

```python
>>> manager = SharedMemoryManager()
>>> manager.start()
>>> shr = manager.SharedMemory(name='foo', size=10)
```

Also, perhaps a naive question: don't you *always* need a `start()`ed manager in order for the processes to communicate? Doesn't `SharedMemoryServer` has to be involved? 

In a more general sense what I'm aiming at here is trying to provide the simplest API possible, meaning exposing less rather than more, especially for low-level details, because anything (function, class, argument, etc.) that is public or documented is basically set into stone.

applio replied:
This topic has also moved to b.p.o. -- I *think* we are covering all aspects there.

giampaolo commented on Doc/library/multiprocessing.shared_memory.rst:
```diff
@@ -0,0 +1,381 @@
+
+   .. attribute:: size
+
+      Read-only access to size in bytes of the shared memory block.
```
Is this the size previously passed as argument or the actual size of the occupied by all memory blocks (aka, the whole object's size)?

applio replied:
It is the requested size in bytes of the current (only one) shared memory block.  It is possible to attach to a larger shared memory block but only request access to the first n bytes.  One use case for this would be to avoid any potential for processes to access rapidly changing parts of the shared memory block (since two threads or processes reading and writing to the same location in memory is dangerous...)

The actual size of the whole shared memory block is obtainable.  When requesting to attach to an existing shared memory segment, supply size=0 and the actual size of the existing shared memory block is discovered and made available.  This is probably the dominant use case when attaching to existing shared memory blocks.

giampaolo commented on Doc/library/multiprocessing.shared_memory.rst:
```diff
@@ -0,0 +1,381 @@
+
+When using a :class:`SharedMemoryManager` in a :keyword:`with` statement, the
+shared memory blocks created using that manager are all released when the
+:keyword:`with` statement's code block finishes execution.
```
Again, having this in `multiprocessing.managers` would allow the repetition of the `with` statement usage/clarification.

applio replied:
In all my years of working with multiprocessing even before becoming a core dev, it has been extraordinarily rare to meet anyone who knew of multiprocessing.managers and even fewer people have used one.  I believe this is because (1) multiprocessing.managers are not commonly needed for the vast majority of use cases, and (2) they are not being discovered because there is too much information and too many non-trivial concepts covered in the multiprocessing main documentation.

When a user wishes to use the zero-copy shared memory functionality, they will very commonly also want to use SharedMemoryManager.  We should help users make this mental connection right away.

giampaolo replied:
I agree the preferred way should be using `SharedMemoryManager` because it's more high level. 

> In all my years of working with multiprocessing even before becoming a core dev, it has been extraordinarily rare to meet anyone who knew of multiprocessing.managers and even fewer people have used one.

Mmm... apparently it's not so rare:
https://www.programcreek.com/python/example/8456/multiprocessing.Manager

giampaolo commented on Lib/multiprocessing/shared_memory.py:
```diff
@@ -8,34 +8,119 @@
+            try:
+                h_map = kernel32.OpenFileMappingW(FILE_MAP_READ, False, name)
+            except OSError as ose:
+                raise ExistentialError(*ose.args)
```
I'd be for letting OSError propagate

applio replied:
I think this makes sense too, per Neil's suggestions.

applio replied:
Thank you Neil for changing this to a ``FileNotFoundError`` in commit a5800a910fd96d113f029754d7b1cb3aee4e22b9.

giampaolo commented on Lib/multiprocessing/shared_memory.py:
```diff
@@ -8,34 +8,119 @@
+            except OSError as ose:
+                raise ExistentialError(*ose.args)
+            p_buf = kernel32.MapViewOfFile(h_map, FILE_MAP_READ, 0, 0, 0)
+            kernel32.CloseHandle(h_map)
```
`CloseHandle` should be in a `finally` block in case `MapViewOfFile` fails

applio replied:
Oof -- you are right.  Thank you!  Fixed in commit 107656716917bd68bded9f746c6b031a01ecaa39.

giampaolo commented on Lib/multiprocessing/shared_memory.py:
```diff
@@ -134,7 +243,7 @@ def __init__(self, *args, buffer=None, **kwargs):
             try:
                 existing_type.__init__(self, *args, **kwargs)
-            except:
+            except Exception:
```
why suppress the error here? (probably deserves a comment)

applio replied:
Agreed this deserves a comment.  Anecdotally (but unfortunately I do not immediately remember which class), one or more classes that I successfully tested this function on had enforced restrictions on what keyword arguments were allowed to be passed into its `__init__` yet extra keyword arguments were permissible/ignored when passed into its `__new__`; for this class, the filling of values into a new object instance was accomplished by its `__new__` and the `__init__` was superfluous for the purposes of the `shareable_wrap` function.  In order to add a comment, I need to find this in my notes but before doing so I would like to advocate for the removal of the `shareable_wrap` function from the module.

As described in the docstring for `shareable_wrap`, its purpose was to make it quick and easy to put objects supporting the buffer protocol as both producer and consumer onto shared memory:
```
>>> import numpy as np
>>> a = np.array([4, 5, 6, 7])
>>> print(a)
[4 5 6 7]
>>> from multiprocessing import shared_memory
>>> shm_a = shared_memory.shareable_wrap(a)
>>> shm_a
ndarrayShareable(shape=(4,), strides=(8,), dtype=dtype('int64'), shmem_name='psm_25371_85606', cls=<class 'numpy.ndarray'>)
>>> print(shm_a)
[4 5 6 7]
>>> import pickle
>>> with open("./arr.pkl", "wb") as f:
...     pickle.dump(shm_a, f)
... 
>>> with open("./arr.pkl", "rb") as f:
...     shm_b = pickle.load(f)  # Load in a different process
... 
>>> shm_b  # Note the same shared memory block name
ndarrayShareable(shape=(4,), strides=(8,), dtype=dtype('int64'), shmem_name='psm_25371_85606', cls=<class 'numpy.ndarray'>)
>>> shm_b[-1] = 999
>>> print(shm_a)
[  4   5   6 999]
>>> len(pickle.dumps(shm_a))  # Note the size of the serialized representation of a small shared numpy array
122
>>> c = np.array(range(50000))
>>> shm_c = shared_memory.shareable_wrap(c)
>>> len(shm_c)
50000
>>> len(pickle.dumps(shm_c))  # Note the size of the much larger shared array's serialized representation
123
>>> len(pickle.dumps(c))
400162
```

I propose that `shareable_wrap` should be removed from the code for three reasons:
(1) We have no examples of data types in the core of Python that are __both__ producers and consumers of the buffer protocol, hence I do not have a way to demonstrate its use without referring to 3rd party library objects like `numpy.ndarray`.
(2) Feedback from others who have tested `shareable_wrap` generally found it easy to use but found the representation (the output from `__repr__`) of the wrapped objects to be surprising. I am unsure of how to solve this sense of surprise in a truly generic way.
(3) The buffer protocol makes no specification at the Python object api level for how to supply the required inputs.  This `shareable_wrap` function is a pure Python function.  As such it took many iterations to adapt this function to handle the widest possible variety of 3rd party library objects, but there are undoubtedly scenarios / new kinds of objects which it will not be able to handle.

applio replied:
I included this function to provoke discussion around its limitations while demonstrating some of what is possible.  This is also why I skipped including it in the documentation for now.

giampaolo replied:
I'm not sure I fully understand the technical intricacies, but I agree `shareable_wrap` should either go or made it private / not-documented.

applio replied:
The consensus seemed to favor removing it.
I have removed it in commit 885592b56f555c03e6444c7de1839c38c4792fb6.

giampaolo commented on Lib/multiprocessing/shared_memory.py:
```diff
@@ -432,10 +546,15 @@ def __setitem__(self, position, value):
         return struct.unpack_from("q", self.shm.buf, 0)[0]
 
+    def __repr__(self):
+        return f'{self.__class__.__name__}({list(self)}, name={self.shm.name!r})'
```
Does `list(self)` prints all the elements in the iterable? If so, it appears it may print too much stuff on screen (and also unnecessarily make a copy in memory).

applio replied:
Yes, it will copy all of the contents into a new list object.  This can be expensive and overwhelm the user in the interactive shell.  This does meet the expectations of` __repr__` though.

I am familiar with some 3rd party library objects cheating and not showing everything necessary for the re-creation of the object (as repr asks us to do).  Do you think we should do something similar?  For example:  ShareableList([1, 2, 3, ...], name='blah')

applio replied:
The more I think about this, the more I like the idea of removing` __repr__` and offering `__str__` instead which, when the length of the list is >1000 (a cutoff employed by numpy), produces output that looks like this:  ShareableList([0, 1, 2, ..., 998, 999, 1000], name='blah')

giampaolo replied:
I'm also still not so sure now. FWIW one may argue it could even `return list(self)`:

```
>>> import multiprocessing.managers
>>> mgr = multiprocessing.managers.SyncManager()
>>> mgr.start()
>>> ll = mgr.list((1,2,3))
>>> ll
[1, 2, 3]
```

The "name" parameter seems important though. I'll leave it up to you.

giampaolo commented on Lib/multiprocessing/shared_memory.py:
```diff
@@ -464,11 +583,13 @@ def _offset_packing_formats(self):
 
-    @classmethod
-    def copy(cls, self):
+    def copy(self, *, name=None):
```
What's the use case of `name`? Copying another `ShareableList` instance? What's the point?

applio replied:
When creating `SharedMemory` instances, the user can control the name of the shared memory block (to support a variety of use cases).  Here too, when creating a copy, the user can specify the name of the new shared memory block about to be created to hold the copy.

giampaolo replied:
IMO this is an unnecessary over-complication of the API. If the user wants [a copy of] another memory region they can do so by instantiating a new `SharedMemory` instance (which really is what you're doing here). I recommend mimicking the `list` object as much as possible (aka `list.copy` doesn't have a "name" parameter).

applio replied:
Agreed -- removed in commit 715ded90abdde274159b219172a844a3a7da8424.

giampaolo commented on Lib/multiprocessing/shared_memory.py:
```diff
@@ -493,18 +614,22 @@ def __init__(self, name, segment_names=[]):
-    def register_segment(self, segment):
-        util.debug(f"Registering segment {segment.name!r} in pid {os.getpid()}")
-        self.segment_names.append(segment.name)
+    def register_segment(self, segment_name):
```
This is a public name but it's not documented. Shouldn't it be private?

applio replied:
I have debated whether `SharedMemoryTracker` should be made private or properly documented as public.

In my own use of this code over the last year, I have found reasons to use `SharedMemoryTracker` independent of its use through `SharedMemoryManager`.  But I am unconvinced that it will find general use and that my use of it is only an edge case.

I hope that someone will react to it during review and say, "this is awesome, why is this not documented properly?!?"  If no one has a strong reaction, then I agree it should be private to the module.

giampaolo replied:
If you consider it an edge case and are unsure then I suggest not documenting this class and possibly make it private (that way the methods name would be OK). To me this class looks more like a low level component of "support" which is only used by the `Server` class.

applio replied:
Changed ``SharedMemoryTracker`` to be ``_SharedMemoryTracker`` in commit 594140a65489adde4f60b14925e295bb1639501e.

giampaolo commented on Lib/multiprocessing/shared_memory.py:
```diff
@@ -493,18 +614,22 @@ def __init__(self, name, segment_names=[]):
+        self.segment_names.append(segment_name)
 
     def destroy_segment(self, segment_name):
+        """Calls unlink() on the shared memory block with the supplied name
```
same here (shouldn't it be private)?

applio replied:
Same comment as on `register_segment`.

giampaolo commented on Lib/multiprocessing/shared_memory.py:
```diff
@@ -540,19 +671,39 @@ def create(self, c, typeid, *args, **kwargs):
+    def list_segments(self, c):
+        """Returns a list of names of shared memory blocks that the Server
+        is currently tracking."""
+        return self.shared_memory_context.segment_names
```
shouldn't these 3 methods be private?

applio replied:
`list_segments()` specifically is useful as a debugging tool.  It can be invoked remotely via `multiprocessing.managers.dispatch()` and it has helped me both troubleshoot and verify things.

If you are suggesting that `SharedMemoryServer` itself should be marked as private, my thinking on this has been split.  I have considered that it should still be added to the documentation for completeness sake because the main multiprocessing documentation describes the `Server` class as being intimately related to using managers.  But I am unable to provide a use case where a user would wish to instantiate their own `SharedMemoryServer`.  I believe the right thing to do is to add a brief entry for it to the documentation...

giampaolo replied:
As per module doc, it seems the "server" is described as something returned by the manager via `Manager().get_server()` but I'm not sure why you may want to access it (and the doc itself doesn't say). So I would not consider it too important as a component (probably it's not even worth documenting it).

applio replied:
This is something lacking in the documentation on mangers.  Admittedly it is difficult to do a good job explaining the uses for low-level controls sometimes but `get_server()` enables low-level control, providing the option to directly control the starting and stopping of the `Server` whereas calling `Manager().start()` asks the manager to do that for you.

Your encouragement that it's not worth documenting `SharedMemoryServer` makes sense.  I believe that we do still need to improve documentation around managers and their use of the `Server` class.

giampaolo commented on Lib/multiprocessing/shared_memory.py:
```diff
@@ -563,11 +714,53 @@ def get_server(self):
+                sms.unlink()
+                raise e
+        finally:
+            conn.close()
```
Could use a `with` statement (`with self.Client(...) as client:`).

applio replied:
Nice! I did not think conn had a context manager in all cases but it does -- cool.  This and its near-mirror in your next comment are now changed in commit 0be05318219ca1872f07af45404aca5619c3f0d5.

giampaolo commented on Lib/multiprocessing/shared_memory.py:
```diff
@@ -563,11 +714,53 @@ def get_server(self):
+                sl.shm.unlink()
+                raise e
+        finally:
+            conn.close()
```
Could use a `with` statement (`with self.Client(...) as client:`).

applio replied:
Changed -- see prior comment's response.

giampaolo commented on Lib/multiprocessing/shared_memory.py:
```diff
@@ -563,11 +714,53 @@ def get_server(self):
+SharedMemoryManager.register('Event', threading.Event, EventProxy)
+SharedMemoryManager.register('Lock', threading.Lock, AcquirerProxy)
+SharedMemoryManager.register('RLock', threading.RLock, AcquirerProxy)
+SharedMemoryManager.register('Semaphore', threading.Semaphore, AcquirerProxy)
```
Since `SharedMemoryManager` is a subclass of `SyncManager` aren't these already registered?

applio replied:
It is no longer a subclass of SyncManager -- it is merely a subclass of BaseManager.  I changed this while writing the docs when I realized how confusing it was to try to explain why Value, Array, list, dict, etc. (distributed shared memory objects) were present on something devoted to non-distributed shared memory.

giampaolo commented on Lib/test/_test_multiprocessing.py:
```diff
@@ -3610,6 +3616,246 @@ def test_copy(self):
+
+    def setUp(self):
+        if not HAS_SHMEM:
+            self.skipTest("requires multiprocessing.shared_memory")
```
You can decorate the class instead

applio replied:
Most of the other tests in `_test_multiprocessing.py` use this (potentially older) style rather than the decorator, but this is no longer universally true.  I will adopt the decorator style for new tests added to the testing module.  Changed in commit 1e5341eb171d7d2b988880020cfcc0a64021326d as well.

giampaolo commented on Lib/test/_test_multiprocessing.py:
```diff
@@ -3610,6 +3616,246 @@ def test_copy(self):
+
+        finally:
+            # Prevent test failures from leading to a dangling segment.
+            sms.unlink()
```
You could use `self.addCleanup(sms.unlink)` right after creating `sms`, spare this long try/finally block and save one indentation.

applio replied:
Good thought.  Changed in commit 1e5341eb171d7d2b988880020cfcc0a64021326d.

giampaolo commented on Lib/test/_test_multiprocessing.py:
```diff
@@ -3610,6 +3616,246 @@ def test_copy(self):
+            sms.close()
+
+        finally:
+            sms.unlink()
```
You could use `self.addCleanup(sms.unlink)` right after creating `sms`, spare this long try/finally block and save one indentation.

applio replied:
Good thought -- same as prior comment.

giampaolo commented on Lib/test/_test_multiprocessing.py:
```diff
@@ -3610,6 +3616,246 @@ def test_copy(self):
+                        sms_uno.close()
+
+                    finally:
+                        sms_uno.unlink()  # A second shm_unlink() call is bad.
```
also here (could use `addCleanup`)

applio replied:
Actually, this one needs to stay as it is, because this second call to `unlink()` is needed to trigger the exception that we are testing for.

giampaolo commented on Lib/test/_test_multiprocessing.py:
```diff
@@ -4873,6 +5103,27 @@ def test_pool(self):
+
+    def test_joinable_queue(self):
+        self.test_queue("JoinableQueue")
+
```
I don't understand why you need to move these 2 methods from the previous tests. The logic appears split between the mixing class and this one for no apparent reason (but I may be missing it). If you need a different logic for certain `TestSharedMemoryManagerTypes` methods/tests you can simply override them (e.g. if you want to skip them).

applio replied:
Both `Queue` and `JoinableQueue`  are implemented as distributed shared memory (a `queue.Queue` object existing in one process and an `AutoProxy` object existing in the local process).  For the same reasons mentioned in my b.p.o. comment, I believe it is much clearer to not include these on `SharedMemoryManager` by default and to instead document how they may be added on-demand with a single line of code (by calling `register()`).

There is also the potential for a POSIX shared memory message queue (not currently implemented in shared_memory.py).  This is also something worth considering for the future.

giampaolo replied:
Question: if we expose them they are not subject to the speedup, correct? Would it be the same as using them via the `SyncManager`?

applio replied:
Correct, exposing `Queue` or `JoinableQueue` here would not make them faster; they would be the same as if using them via `SyncManager`.

We could simply encourage users to employ a `SyncManager` anytime they want to use a `Queue` or `Lock` but this forces the creation of two managers with a server behind each and therefore two processes instead of only one.  I think it will be easier to explain, "create one manager to oversee work and provide cleanup at the end."

Although my use of `empty()` in this example is admittedly fragile, this example shows augmenting `SyncManager` with `Queue` is really only one line of code:
```
from multiprocessing import shared_memory
import queue
shared_memory.SharedMemoryManager.register('Queue', queue.Queue)
with shared_memory.SharedMemoryManager() as smm:
    sl = smm.ShareableList("feb")  # fast, zero-copy shared memory
    q = smm.Queue()                # slow, copy+message distributed shared memory
    for letter in sl:
        q.put(letter.upper())
    while not q.empty():
        print(q.get(timeout=1))
```

giampaolo replied:
I also think we should get rid of Queue, Lock, etc. Just for confirmation:  `ShareableList` and `AsyncManager.Lock` can be safely mixed/used together, correct?

applio replied:
Agreed -- I have now gotten rid of Queue, Lock, etc. in commit 395709bbfa564c528d6d8aeba780d988bb774ede.

Yes, confirmed:  ``ShareableList`` and ``SyncManager.Lock`` can be safely mixed/used together.

I was sure you meant ``SyncManager`` when your wrote ``AsyncManager`` but I still went searching just to make sure there was not some ``AsyncManager`` hiding inside ``asyncio`` somewhere that I was not thinking of.  =)

applio committed:
Added Sphinx doctest run controls.

applio committed:
CloseHandle should be in a finally block in case MapViewOfFile fails.

applio committed:
Missed opportunity to use with statement.

applio committed:
Switch to self.addCleanup to spare long try/finally blocks and save one indentation, change to use decorator to skip test instead.

nascheme committed:
Simplify the posixshmem extension module.

Provide shm_open() and shm_unlink() functions.  Move other
functionality into the shared_memory.py module.

giampaolo commented on Doc/library/multiprocessing.shared_memory.rst:
```diff
@@ -0,0 +1,388 @@
+   ...     sl = smm.ShareableList(range(2000))
+   ...     # Divide the work among two processes, storing partial results in sl
+   ...     p1 = Process(target=do_work, args=(sl.shm.name, 0, 1000))
+   ...     p2 = Process(target=do_work, args=(sl.shm.name, 1000, 2000))
```
Instead of passing the name to the worker function, why not pass the `ShareableList` instance instead? 

```python
def do_work(shareable_list):
    print(shareable_list)
    shareable_list[0] = 30


with shared_memory.SharedMemoryManager() as smm:
    mem = smm.SharedMemory(size=10)
    sl = smm.ShareableList(range(10))
    proc = Process(target=do_work, args=(sl, ))
    proc.start()
    proc.join()
    print(sl)  # index 0 is 30
```

Also, that would be consistent with `SyncManager` usage when it comes to using it's objects (list, Lock, etc... they are all passed around as-is).

applio replied:
Agreed -- this makes sense and should be added.

applio replied:
Added in commit 5848ec413bee0e9f5dd068f5d7563ac7786e329c.

giampaolo commented on Lib/multiprocessing/shared_memory.py:
```diff
@@ -53,13 +140,21 @@ def unlink(self):
+                read_only=read_only
+            )
 
         self._mmap = mmap.mmap(self.fd, self.size)
```
This is not my territory, so take it with a grain of salt. I see that `mmap.mmap` provides some interesting arguments such as `access`, `prot`(ection), and `flags`:
https://docs.python.org/3/library/mmap.html#mmap.mmap
Depending on `mode` and `read_only` args that you provide here perhaps it would make sense to specify parameters for `mmap` in accordance (e.g. `flags=MAP_SHARED`, `prot=PROT_WRITE/READ` and `access=ACCESS_READ/WRITE`).
This is a private argument so *maybe* it's not so important. Just mentioning it for completeness/correctness.

applio replied:
It is a good point.  With the simpler API introduced in commit 0d3d06f7cbcb47b2dee6c027c35265795052c899, we constrain the supported behaviors to no longer offer these sorts of control.  The behavior of mmap seems to have some inconsistencies of its own across different platforms when trying to specify such modes of access -- I think this only encourages me to believe we have made the right choice with the simplified API.

giampaolo commented on Lib/multiprocessing/shared_memory.py:
```diff
@@ -8,34 +8,121 @@
+                    f"Shared memory already exists with name={name}"
+                )
+
         self._mmap = mmap.mmap(-1, size, tagname=name)
```
Same here (see my previous [comment](https://github.com/python/cpython/pull/11816/files#r257460172)).

applio replied:
Same comment for me too (ref the prior comment/conversation).

giampaolo commented on Lib/test/_test_multiprocessing.py:
```diff
@@ -3610,6 +3616,234 @@ def test_copy(self):
 
+
+@unittest.skipUnless(HAS_SHMEM, "requires multiprocessing.shared_memory")
+class _TestSharedMemory(BaseTestCase):
```
is there a reason for this to be this private? (underscore)

giampaolo replied:
It seems this is class is not executed. Looks like you want to inherit from unittest.TestCase.

applio replied:
It is executing -- I see it in the tests being run when I enable `-v`.  I see it running on the buildbots too (more reassuring to me).

The pattern used by this class is being used on a number of others and is motivated/related to the controls for when a test runs under fork/forkserver/spawn.

giampaolo replied:
It doesn't run with:

```
./python -m test _test_multiprocessing -v -m _TestSharedMemory
```

applio replied:
Correct.  Instead something dynamic happens to the name.  Check out what happens to the name:
```
test_shared_memory_ShareableList_basics (test.test_multiprocessing_spawn.WithProcessesTestSharedMemory) ... ok
test_shared_memory_ShareableList_pickling (test.test_multiprocessing_spawn.WithProcessesTestSharedMemory) ... ok
test_shared_memory_SharedMemoryManager_basics (test.test_multiprocessing_spawn.WithProcessesTestSharedMemory) ... ok
test_shared_memory_across_processes (test.test_multiprocessing_spawn.WithProcessesTestSharedMemory) ... ok
test_shared_memory_basics (test.test_multiprocessing_spawn.WithProcessesTestSharedMemory) ... ok
```

This is due to ``ALLOWED_TYPES = ('processes',)`` appearing at the head of the test class.

giampaolo replied:
It appears `BaseTestCase` does not inherit from `unittest.TestCase`, that's why it's not possible to run this test individually.

applio committed:
Added to doc around size parameter of SharedMemory.

applio committed:
Changed PosixSharedMemory.size to use os.fstat.

applio committed:
Change SharedMemory.buf to a read-only property as well as NamedSharedMemory.size.

applio committed:
Marked as provisional per PEP411 in docstring.

applio committed:
Merge branch 'enh-tests-neilsimplify-shmem' into enh-tests-shmem
adopting Neil's good work centralizing name autogeneration and removing
custom exceptions.

applio committed:
Changed SharedMemoryTracker to be private.

applio committed:
Removed registered Proxy Objects from SharedMemoryManager.

applio committed:
Removed shareable_wrap().

applio committed:
Removed shareable_wrap() and dangling references to it.

applio committed:
Merge remote and local branches regarding elimination of
shareable_wrap().

giampaolo commented on Lib/multiprocessing/shared_memory.py:
```diff
@@ -115,57 +113,146 @@ def __init__(self, name, flags=None, mode=384, size=0, read_only=False):
-                raise ExistentialError(
-                    f"Shared memory already exists with name={name}"
-                )
+                raise FileExistsError(name)
```
You could enrich the exception:

```python
raise FileExistsError(errno.EEXIST, os.strerror(errno.EEXIST), name)
```

Result:

```diff 
- FileExistsError: 'filename'
+ FileExistsError: [Errno 17] File exists: 'filename'
```

applio replied:
Good point -- changed in 68785333a25d902a42252ace3a8268357efd6094.

giampaolo commented on Lib/multiprocessing/shared_memory.py:
```diff
@@ -89,15 +87,15 @@ class WindowsNamedSharedMemory:
-            except OSError as ose:
-                raise ExistentialError(*ose.args)
+            except OSError:
+                raise FileNotFoundError(name)
```
This is too broad. Suggestion: look for specific errno (not tested):


```python
try:
    h_map = kernel32.OpenFileMappingW(FILE_MAP_READ, False, name)
except OSError as err:
    if err.errno == errno.ENOENT:
        raise FileExistsError(errno.ENOENT, os.strerror(errno.ENOENT), name)
    raise
```

eryksun replied:
The try/except shouldn't be necessary anymore, depending on the code in _winapi.

applio replied:
Good point -- it was nice to eliminate the try/except in commit 68785333a25d902a42252ace3a8268357efd6094.

giampaolo commented on Lib/multiprocessing/shared_memory.py:
```diff
@@ -115,57 +113,146 @@ def __init__(self, name, flags=None, mode=384, size=0, read_only=False):
+        if self.fd >= 0:
+            return os.fstat(self.fd).st_size
+        else:
+            return 0
```
Mmm... since you make a `os.fstat` call this may fail. And it's kinda unusual for a `getattr` to fail. Also, AFAIU the size may change if the memory is enlarged (it can happen I suppose). As such I think it may make sense to provide this as a method, not a property.

applio replied:
In commit 0d3d06f7cbcb47b2dee6c027c35265795052c899, I have eliminated the use of ``os.fstat()`` in the property for ``size``.  Instead, it is now an unchanging value stored as part of the ``__init__()``.

giampaolo commented on Lib/multiprocessing/shared_memory.py:
```diff
@@ -115,57 +113,146 @@ def __init__(self, name, flags=None, mode=384, size=0, read_only=False):
+            except FileExistsError:
+                continue
+            self.name = name
+            break
```
Not sure if it's worth it, just mentioning. `tempfile` module provides a max number of retries strategy you may want to reuse here:
https://github.com/python/cpython/blob/825aab95fde959541859383f8ea7e7854ebfd49f/Lib/tempfile.py#L254

giampaolo commented on Lib/multiprocessing/shared_memory.py:
```diff
@@ -115,57 +113,146 @@ def __init__(self, name, flags=None, mode=384, size=0, read_only=False):
 
+    def unlink(self):
+        if self.name:
+            _posixshmem.shm_unlink(self.name)
```
shouldn't `unlink` also call `close`?

applio replied:
In the same sense that ``close()`` should not call ``unlink()``, the reverse should also be true to maintain a clean separation of duties.  This keeps things clear in situations like the following: if the coordinating process is done accessing the shared memory block, it should be permitted to call ``close`` at that time and wait to call ``unlink`` once it knows all other processes are similarly done; in this case we would not want ``unlink`` to call ``close`` a second time.  After all, if ``unlink`` did call ``close``, it would only be able to do so in one process but other processes should still call ``close``.  Best to keep them separated.  There is a reason why this is the design for the shared memory C APIs that we use on various operating systems.

giampaolo replied:
Clear enough. Thanks.

giampaolo replied:
On a second thought... if `unlink()` is called, isn't the instance supposed to be unusable? AFAICT you can still play with the memoryview object but only locally (no data is shared). Also, I wonder whether it makes sense to protect `buf` attribute so that it raises an exception if used after `close()` and `unlink()`.

applio replied:
If one process calls ``unlink()``, we might protect ``buf`` and even do all the other things that ``close()`` would do for that one process but all other processes still need to call ``close()`` themselves.  I think the clean separation of duties is still the right answer here.

The docs currently say that if ``unlink()`` is called, any continued use is unsupported.  The exact behavior of ``unlink`` with respect to when the shared memory block truly is destroyed is not something we can ultimately control or enforce.  Users must call ``close()`` and ``unlink()`` to ensure proper cleanup.  If users can not follow this directive, I doubt we could ever truly protect them from themselves.

applio committed:
For consistency added __reduce__ to key classes.

applio committed:
Fix for potential race condition on Windows for O_CREX.

applio committed:
Remove unused imports.

osvenskan commented on Lib/multiprocessing/shared_memory.py:
```diff
@@ -138,34 +130,99 @@ def unlink(self):
 
-class PosixSharedMemory(_PosixSharedMemory):
+# FreeBSD (and perhaps other BSDs) limit names to 14 characters.
+_SHM_SAFE_NAME_LENGTH = 14
```
I don't know if this limit bothers anyone, but it's probably not relevant anymore. I wrote `posix_ipc` back when FreeBSD was still developing POSIX IPC support, so in the early days (~8-10 years ago) it was still a little rough. That's probably where this limit came from. It improved a lot between FreeBSD 7.2 and 8.0. I just tested FreeBSD 11.1 with my `posix_ipc` module and I was able to create a semaphore with a 91 character name, so somewhere along the line this limit became irrelevant, I just can't say when.

applio replied:
I think I would prefer to err on the side of caution by keeping this for now and potentially relaxing it in the future.  I suspect there are limits on various platforms that still need to be respected but this limit was the most conservative I found documented.  It is perhaps overly optimistic to say so but I hope that this limit (as it stands) does not create any true hardships for users when they try to come up with unique names for things.

giampaolo replied:
FWIW this limit still exists and affects at least Linux, Solaris and AIX. I have code in psutil which takes this into account.

osvenskan reviewed (commented):
Nice simplifications! This looks good to me.

eryksun commented on Lib/multiprocessing/shared_memory.py:
```diff
@@ -1,225 +1,316 @@
+    import ctypes
+    from ctypes import wintypes
+
+    kernel32 = ctypes.windll.kernel32
```
This should be `kernel32 = ctypes.WinDLL('kernel32', use_last_error=True)`. With `ctypes.windll`, it's using a globally cached `WinDLL` instance. This in turn caches function pointers, which leads to conflicting prototypes between packages. Also, `use_last_error` is required for `ctypes.get_last_error()`, which is more reliable than FFI-based `ctypes.GetLastError()`.

Is this ctypes usage just temporary for experimenting until the API is finalized? Ultimately I hope this uses _winapi instead of ctypes.

applio replied:
Thanks, eryksun -- you rock.  Changed both in commit 868b83d681c3e11dd6de22631ac61493a5b73ded.

I believe the API is close enough to being finalized that we can move this to _winapi now.  I would very much appreciate your help reviewing that move, if I can talk you into that.

applio replied:
I have moved this over to use _winapi now, changed in commit 9d83b060b6f0e0df82a8f2fe4833495bf1830181.

If you are up for helping review this too, it would again be very much appreciated.

eryksun commented on Lib/multiprocessing/shared_memory.py:
```diff
@@ -1,225 +1,316 @@
+                name
+            )
+            try:
+                last_error_code = kernel32.GetLastError()
```
This should call `ctypes.get_last_error()`, which is being called elsewhere. But for it to work, we need `use_last_error=True`, as mentioned above.

eryksun replied:
Strike that. `kernel32.CreateFileMappingW._errcheck`  is set  to `_errcheck_bool`, which should raise `FileExistsError` here. Maybe this was added due to a misunderstanding. Using `ctypes.windll.kernel32` renders the `_errcheck` function useless, since `ctypes.get_last_error` isn't enabled.

applio replied:
When ``kernel32.CreateFileMappingW`` is passed the name of an existing anonymous memory block in this way, it succeeds and it does not raise ``FileExistsError`` (``_errcheck_bool`` receives a non-zero ``result`` value).  This motivates the ``ctypes.get_last_error()`` call to check for a return value of ``183``.

This seems to match what is described on docs.microsoft.com, regarding the return value from ``CreateFileMappingA``:
```
    If the object exists before the function call, the function
    returns a handle to the existing object (with its current
    size, not the specified size), and GetLastError returns
    ERROR_ALREADY_EXISTS.
```

FWIW, this technique is leveraged in the Boost library for this same purpose.  The following is a snippet from boost/interprocess/windows_shared_memory.hpp where ``create_file_mapping`` is a wrapper for ``CreateFileMappingW``:
```
   switch(type){
      case ipcdetail::DoOpen:
         m_handle = winapi::open_file_mapping(map_access, filename);
      break;
      case ipcdetail::DoCreate:
      case ipcdetail::DoOpenOrCreate:
      {
         m_handle = winapi::create_file_mapping
            ( winapi::invalid_handle_value, protection, size, filename
            , (winapi::interprocess_security_attributes*)perm.get_permissions());
      }
      break;
      default:
         {
            error_info err = other_error;
            throw interprocess_exception(err);
         }
   }

   if(!m_handle || (type == ipcdetail::DoCreate && winapi::get_last_error() == winapi::error_already_exists)){
      error_info err = system_error_code();
      this->priv_close();
      throw interprocess_exception(err);
   }
```

I would be happier if ``CreateFileMappingW`` behaved as you suggested it might.

eryksun replied:
> it succeeds and it does not raise `FileExistsError` (`_errcheck_bool` receives a non-zero `result` value). This motivates the `ctypes.get_last_error()` call to check for a return value of `183`.

Sorry it slipped my mind that Windows normally uses the [`OBJ_OPENIF`](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/content/wudfwdm/ns-wudfwdm-_object_attributes) disposition when creating a named kernel object. If this flag weren't used, then the underlying `NtCreateSection` system call would fail with `STATUS_OBJECT_NAME_COLLISION`, which maps to Windows `ERROR_ALREADY_EXISTS`. Instead it succeeds, and `CreateFileMapping` manually sets the last error. It's one of a small number of cases in which a WINAPI call succeeds with an error set.

eryksun commented on Lib/multiprocessing/shared_memory.py:
```diff
@@ -1,225 +1,316 @@
+    PAGE_READONLY = 0x02
+    PAGE_EXECUTE_READWRITE = 0x04
+    INVALID_HANDLE_VALUE = -1
+    FILE_ALREADY_EXISTS = 183
```
This constant shouldn't be necessary once other mistakes are corrected. But it should be named `ERROR_ALREADY_EXISTS` if it stays. It's confusing to readers if we invent our own names for well-known error codes.

applio replied:
Thanks for catching this detail about the appropriate name.  Changed in commit 868b83d681c3e11dd6de22631ac61493a5b73ded (at least until it is no longer needed).

applio replied:
These have now been (re-)moved to _winapi per your later suggestion as part of commit 68785333a25d902a42252ace3a8268357efd6094.

applio committed:
Update access to kernel32 on Windows per feedback from eryksun.

applio committed:
Moved kernel32 calls to _winapi.

eryksun commented on Lib/multiprocessing/shared_memory.py:
```diff
@@ -1,225 +1,250 @@
+            try:
+                last_error_code = _winapi.GetLastError()
+                if last_error_code == _winapi.ERROR_ALREADY_EXISTS:
+                    raise FileExistsError(f"File exists: {name!r}")
```
It would be nice to include the `errno` and `winerror` values for completeness: `FileExistsError(errno.EEXIST, os.strerror(errno.EEXIST), name, _winapi.ERROR_ALREADY_EXISTS)`.

applio replied:
Good point -- added in commit 68785333a25d902a42252ace3a8268357efd6094.

eryksun commented on Lib/multiprocessing/shared_memory.py:
```diff
@@ -1,225 +1,250 @@
+            # Attempt to dynamically determine the existing named shared
+            # memory block's size which is likely a multiple of mmap.PAGESIZE.
+            try:
+                h_map = _winapi.OpenFileMappingW(PAGE_READONLY, False, name)
```
This should request `FILE_MAP_READ` (4) access. `PAGE_READONLY` (2) is a page protection for `CreateFileMapping`. It happens to be the same constant value as `FILE_MAP_WRITE` access. 

`CreateFileMapping` determines the desired access for the section object from the given page protection. OTOH, with `OpenFileMapping` we have to supply the desired access explicitly.

applio replied:
Thanks for catching this.  Changed to ``_winapi.FILE_MAP_READ`` in commit 68785333a25d902a42252ace3a8268357efd6094.

eryksun commented on Lib/multiprocessing/shared_memory.py:
```diff
@@ -1,225 +1,250 @@
+            except OSError:
+                raise FileNotFoundError(name)
+            try:
+                p_buf = _winapi.MapViewOfFile(h_map, PAGE_READONLY, 0, 0, 0)
```
This should request `FILE_MAP_READ` (4) access.

It's the other way around in the underlying NT API `NtMapViewOfSection`. It takes the desired page protection and determines the required section access from the protection. So `MapViewOfFile[Ex]` actually reverses this to get the page protection in order to call `NtMapViewOfSection`.

applio replied:
Also changed to ``_winapi.FILE_MAP_READ`` in commit 68785333a25d902a42252ace3a8268357efd6094.

eryksun commented on Lib/multiprocessing/shared_memory.py:
```diff
@@ -1,225 +1,250 @@
+            h_map = _winapi.CreateFileMappingW(
+                INVALID_HANDLE_VALUE,
+                _winapi.NULL,
+                PAGE_EXECUTE_READWRITE,
```
Why do use `PAGE_EXECUTE_READWRITE` protection here? mmap defaults to `PAGE_READWRITE`.

applio replied:
I remember mentally debating this choice but not the details.  Your point that mmap defaults to ``PAGE_READWRITE`` should have dictated what was used but I somehow missed it.

Good catch -- changed to ``_winapi.PAGE_READWRITE`` in commit 68785333a25d902a42252ace3a8268357efd6094.

eryksun commented on Modules/_winapi.c:
```diff
@@ -464,6 +474,41 @@ _winapi_CreateFile_impl(PyObject *module, LPCTSTR file_name,
+    Py_END_ALLOW_THREADS
+
+    if (handle == NULL) {
+        PyErr_SetFromWindowsErr(0);
```
Consider using `PyErr_SetFromWindowsErrWithUnicodeFilename(0, name)`.

applio replied:
Done -- changed in commit 68785333a25d902a42252ace3a8268357efd6094.

eryksun commented on Modules/_winapi.c:
```diff
@@ -464,6 +474,41 @@ _winapi_CreateFile_impl(PyObject *module, LPCTSTR file_name,
 }
 
+/*[clinic input]
+_winapi.CreateFileMappingW -> HANDLE
```
Personally, I'm not against using the real name with the "W" suffix. But thus far this module has used the macro names such as `_winapi.CreateFileMapping`.

Actually, we don't define `UNICODE`, and we have a few functions in _winapi that link to the ANSI version and use `LPCTSTR` -> `LPCSTR` (i.e. `const char *`). This includes `_winapi.CreateFile`. It should instead call `CreateFileW` and use an `LPCWSTR` name.

applio replied:
Thanks for making a point of this.  I would prefer to stick with the established naming pattern unless there was a real reason otherwise.  I have changed it to ``_winapi.CreateFileMapping`` as you suggest and likewise dropped the "W" on ``_winapi.OpenFileMapping`` in commit 68785333a25d902a42252ace3a8268357efd6094.

eryksun commented on Lib/multiprocessing/shared_memory.py:
```diff
@@ -1,225 +1,250 @@
+
+    PAGE_READONLY = 0x02
+    PAGE_EXECUTE_READWRITE = 0x04
+    INVALID_HANDLE_VALUE = -1
```
I'd  add `INVALID_HANDLE_VALUE` and the memory API  constants to _winapi. 

```
WINAPI_CONSTANT(F_HANDLE, INVALID_HANDLE_VALUE);
WINAPI_CONSTANT(F_DWORD, FILE_MAP_READ);
WINAPI_CONSTANT(F_DWORD, PAGE_READWRITE);
```

Maybe add the rest, too, because why not.

```
WINAPI_CONSTANT(F_DWORD, FILE_MAP_COPY);
WINAPI_CONSTANT(F_DWORD, FILE_MAP_WRITE);
WINAPI_CONSTANT(F_DWORD, FILE_MAP_EXECUTE);
WINAPI_CONSTANT(F_DWORD, FILE_MAP_ALL_ACCESS);
WINAPI_CONSTANT(F_DWORD, SEC_IMAGE);
WINAPI_CONSTANT(F_DWORD, SEC_RESERVE);
WINAPI_CONSTANT(F_DWORD, SEC_COMMIT);
WINAPI_CONSTANT(F_DWORD, SEC_NOCACHE);
WINAPI_CONSTANT(F_DWORD, SEC_WRITECOMBINE);
WINAPI_CONSTANT(F_DWORD, SEC_LARGE_PAGES);
WINAPI_CONSTANT(F_DWORD, PAGE_NOACCESS);
WINAPI_CONSTANT(F_DWORD, PAGE_READONLY);
WINAPI_CONSTANT(F_DWORD, PAGE_WRITECOPY);
WINAPI_CONSTANT(F_DWORD, PAGE_EXECUTE);
WINAPI_CONSTANT(F_DWORD, PAGE_EXECUTE_READ);
WINAPI_CONSTANT(F_DWORD, PAGE_EXECUTE_READWRITE);
WINAPI_CONSTANT(F_DWORD, PAGE_EXECUTE_WRITECOPY);
WINAPI_CONSTANT(F_DWORD, PAGE_GUARD);
WINAPI_CONSTANT(F_DWORD, PAGE_NOCACHE);
WINAPI_CONSTANT(F_DWORD, PAGE_WRITECOMBINE);
WINAPI_CONSTANT(F_DWORD, MEM_COMMIT);
WINAPI_CONSTANT(F_DWORD, MEM_RESERVE);
WINAPI_CONSTANT(F_DWORD, MEM_FREE);
WINAPI_CONSTANT(F_DWORD, MEM_PRIVATE);
WINAPI_CONSTANT(F_DWORD, MEM_MAPPED);
WINAPI_CONSTANT(F_DWORD, MEM_IMAGE);
```

applio replied:
Sounds plenty good to me -- all have been added to _winapi in commit 68785333a25d902a42252ace3a8268357efd6094.  Thanks!

applio committed:
Removed ShareableList.copy as redundant.

applio committed:
Changes to _winapi use from eryksun feedback.

applio committed:
Adopt simpler SharedMemory API, collapsing PosixSharedMemory and WindowsNamedSharedMemory into one.

applio committed:
Fix missing docstring on class, add test for ignoring size when attaching.

applio committed:
Moved SharedMemoryManager to managers module, tweak to fragile test.

eryksun commented on Modules/_winapi.c:
```diff
@@ -1295,6 +1340,64 @@ _winapi_GetVersion_impl(PyObject *module)
+    Py_END_ALLOW_THREADS
+
+    if (handle == NULL) {
+        PyErr_SetFromWindowsErr(0);
```
Use `PyErr_SetFromWindowsErrWithUnicodeFilename(0, name)`.

applio replied:
Changed in commit caf0a5d9cdfd569c438ce85ece952eacdf4100b8.

I should have thought to change this when you suggested similar on CreateFileMapping.  Thanks!

applio committed:
Tweak to exception in OpenFileMapping suggested by eryksun.

giampaolo commented on Modules/_winapi.c:
```diff
@@ -464,6 +474,41 @@ _winapi_CreateFile_impl(PyObject *module, LPCTSTR file_name,
+        handle = INVALID_HANDLE_VALUE;
+    }
+
+    return handle;
```
Shouldn't you `return NULL` instead of `INVALID_HANDLE_VALUE` in case of error? It appears this way you'll return an int instead of causing an exception.

eryksun replied:
[`HANDLE_return_converter`](https://github.com/python/cpython/blob/caf0a5d9cdfd569c438ce85ece952eacdf4100b8/Modules/_winapi.c#L171) checks for `INVALID_HANDLE_VALUE` with an error set. For `NULL` it returns `None`, which supports `GetStdHandle` since a standard-handle value may be `NULL`. Also, since an error must also be set, `GetCurrentProcess` works with the same converter even though the value it returns is equivalent to `INVALID_HANDLE_VALUE`.

It's  a bit confusing if you're not aware of how it's implemented. It would probably be better to have a special handle type for functions that fail with `INVALID_HANDLE_VALUE`, since by far the majority of WINAPI functions that return a handle (directly or as an out parameter) either use `NULL` or `FALSE` to indicate an error.

applio replied:
@eryksun: Would it make more sense to handle the change you're suggesting by opening a separate issue and PR for it?

I'm hesitant to make such a "code maintenance" change impacting unrelated code as part of this PR which is an "enhancement".

@eryksun: I will offer to help supply the patch if you will review it.

eryksun replied:
It's a minor change, but @zooba should approve it in a new issue before going forward with a PR. I'd rename the current type to `HANDLE_I` and add a new `HANDLE` type that uses `NULL` for errors.

Another maintenance issue is removing the use of the `TCHAR` API and unintentional linking with ANSI functions such as `CreateFileA`, which I think would be fine to fix here if you want. The internal use of the affected functions hasn't had a problem thus far because they use ASCII strings, e.g. for pipe names, but it may eventually cause an issue.

applio replied:
Let's capture all of that in a new issue on b.p.o.

giampaolo commented on Modules/_winapi.c:
```diff
@@ -1295,6 +1340,64 @@ _winapi_GetVersion_impl(PyObject *module)
+        handle = INVALID_HANDLE_VALUE;
+    }
+
+    return handle;
```
same here

giampaolo commented on Modules/_winapi.c:
```diff
@@ -1490,6 +1593,32 @@ _winapi_TerminateProcess_impl(PyObject *module, HANDLE handle,
+    Py_END_ALLOW_THREADS
+
+    if (size_of_buf == 0)
+        PyErr_SetFromWindowsErr(0);
```
and here

giampaolo commented on Lib/multiprocessing/shared_memory.py:
```diff
@@ -1,225 +1,231 @@
+            _posixshmem.shm_unlink(self.name)
 
 
 encoding = "utf8"
```
This is only used by `ShareableList`. Why not make it a class attribute of that class instead? Also not sure if there are use-cases where one may want to change this (if not it's probably better to set it private).

applio replied:
Though it did not get included, the plan was to also supply a ``ShareableDict`` and ``encoding`` was to be used across both.  Anticipating that a ``ShareableDict`` will still land in the future, let's keep it outside the class but I agree it makes sense to make this private.

Changed to private in commit 12c097d88e7d1ceb159a1b726f229654e6218eed.

giampaolo commented on Lib/multiprocessing/shared_memory.py:
```diff
@@ -1,225 +1,231 @@
 
 
-class PosixSharedMemory(_PosixSharedMemory):
+O_CREX = os.O_CREAT | os.O_EXCL
```
Consider making this private (`_O_CREX`)

applio replied:
Good point.  Changed to private in commit 12c097d88e7d1ceb159a1b726f229654e6218eed.

giampaolo commented on Modules/_multiprocessing/posixshmem.c:
```diff
@@ -1,659 +1,112 @@
+
+    if (fd < 0) {
+        if (!async_err)
+            PyErr_SetFromErrnoWithFilenameObject(PyExc_OSError, path);
```
looks like you want to `Py_DECREF(path)` here?

applio replied:
Perhaps I am missing something...  It looks like ``PyUnicode_AsUTF8`` does not create a new reference so there is nothing to decrement?

giampaolo replied:
Sorry, you are right.

giampaolo commented on Modules/_multiprocessing/posixshmem.c:
```diff
@@ -1,659 +1,112 @@
+
+    if (rv < 0) {
+        if (!async_err)
+            PyErr_SetFromErrnoWithFilenameObject(PyExc_OSError, path);
```
looks like you want to `Py_DECREF(path)` here?

applio replied:
The same as my prior comment.

Also, this code is definitely exercised in the tests but I am not seeing any refleaks when running the tests.

giampaolo reviewed (changes requested):
Much better overall. I added some inline comments.

applio committed:
Mark a few dangling bits as private as suggested by Giampaolo.

jameshamm commented on Lib/multiprocessing/shared_memory.py:
```diff
@@ -1,228 +1,234 @@
+    "Create a random filename for the shared memory object."
+    # number of random bytes to use for name
+    nbytes = (_SHM_SAFE_NAME_LENGTH - len(_SHM_NAME_PREFIX)) // 2
+    assert nbytes >= 2, '_SHM_NAME_PREFIX too long'
```
This assert seems conservative. Given a prefix of length 11 or 12 (assuming a 14 character limit), nbytes would be 1, and you could still make 256 valid filenames. Is that enough? Could that number of filenames be exhausted?

applio replied:
The choice to have any prefix at all is largely down to a desire to have some sense of human-recognizable pattern to the auto-generated names but there is no need or motivation to make such a long prefix.  Users can specify their own choice of name for a shared memory block.  This auto-generated name mechanism is purely for convenience so that users do not need to do so themselves.

jakirkham commented on Lib/multiprocessing/managers.py:
```diff
@@ -8,7 +8,8 @@
 
-__all__ = [ 'BaseManager', 'SyncManager', 'BaseProxy', 'Token' ]
+__all__ = [ 'BaseManager', 'SyncManager', 'BaseProxy', 'Token',
+            'SharedMemoryManager' ]
```
According to [this comment]( https://bugs.python.org/issue35813#msg338180 ), sounds like adding `SharedMemoryManager` to `__all__` should be handled to the `try...except` below.

PS: Nice work on this btw!
